File: frontend/src/schemas/ResidentSchema.ts
import { z } from "zod";

export const residentSchema = z.object({
  residence_id: z.string().min(1, "Selecione uma residência"),

  // --- IDENTIFICAÇÃO ---
  nome_completo: z.string().min(3, "Nome obrigatório"),
  nome_social: z.string().optional(),
  cpf: z.string().optional(), // Opcional pois pode não ter
  cns: z.string().optional(),
  data_nascimento: z.string().min(1, "Data obrigatória"),
  sexo: z.enum(["masculino", "feminino"]),
  raca_cor: z.enum(["branca", "preta", "parda", "amarela", "indigena"]),
  etnia: z.string().optional(), // Apenas se indígena
  nis_pis_pasep: z.string().optional(),

  // --- FILIAÇÃO ---
  nome_mae: z.string().min(1, "Nome da mãe é obrigatório (ou marque desconhecida)"),
  mae_desconhecida: z.boolean().default(false),
  nome_pai: z.string().optional(),
  pai_desconhecido: z.boolean().default(false),

  // --- NACIONALIDADE ---
  nacionalidade: z.enum(["brasileira", "naturalizado", "estrangeiro"]),
  pais_nascimento: z.string().optional(),
  dt_naturalizacao: z.string().optional(),
  portaria_naturalizacao: z.string().optional(),
  municipio_nascimento: z.string().optional(),

  // --- CONTATO ---
  telefone_celular: z.string().optional(),
  email: z.string().email("E-mail inválido").optional().or(z.literal("")),

  // --- RESPONSABILIDADE FAMILIAR ---
  eh_responsavel_familiar: z.boolean().default(false),
  cns_cpf_responsavel: z.string().optional(),
  relacao_parentesco: z.enum([
    "conjuge_companheiro", "filho", "enteado", "neto_bisneto", "pai_mae", 
    "sogro", "irmao_irma", "genro_nora", "outro_parente", "nao_parente"
  ]).optional(),

  // --- SOCIODEMOGRÁFICO ---
  ocupacao: z.string().optional(),
  frequenta_escola: z.boolean().default(false),
  escolaridade: z.enum([
    "creche", 
    "pre_escola", 
    "classe_alfabetizacao", 
    "fundamental_1_4", 
    "fundamental_5_8", 
    "fundamental_completo", 
    "fundamental_especial", 
    "eja_iniciais", 
    "eja_finais", 
    "medio", 
    "medio_especial", 
    "medio_eja", 
    "superior", 
    "alfabetizacao_adultos", 
    "nenhum"
  ]).optional(),
  
  situacao_mercado: z.enum([
    "empregador", "assalariado_com_carteira", "assalariado_sem_carteira",
    "autonomo_com_previdencia", "autonomo_sem_previdencia", "aposentado_pensionista", 
    "desempregado", "nao_trabalha", "servidor_publico", "outro"
  ]).optional(),

  responsavel_crianca: z.string().optional(), // Array transformado em string
  frequenta_cuidador: z.boolean().default(false),
  participa_grupos_comunitarios: z.boolean().default(false),
  possui_plano_saude: z.boolean().default(false),
  
  membro_povo_tradicional: z.boolean().default(false),
  povo_tradicional_desc: z.string().optional(),

  // --- IDENTIDADE & DEFICIÊNCIA ---
  orientacao_sexual: z.enum(["heterossexual", "gay", "lesbica", "bissexual", "outro"]).optional(),
  identidade_genero: z.enum(["homem_transexual", "mulher_transexual", "travesti", "outro"]).optional(),
  
  possui_deficiencia: z.boolean().default(false),
  deficiencias_tipo: z.string().optional(), // Array transformado em string

  // --- SAÍDA ---
  saida_cadastro: z.boolean().default(false),
  motivo_saida: z.string().optional(),
  data_obito: z.string().optional(),
  numero_do: z.string().optional()
});

export type ResidentFormValues = z.infer<typeof residentSchema>;

File: backend/app/schemas/resident_schema.py
from typing import Any, List, Optional
from datetime import date, datetime
from pydantic import BaseModel, EmailStr, ValidationInfo, field_validator

from app.models.enums import EscolaridadeEnum, GrauParentescoEnum, IdentidadeGeneroEnum, NacionalidadeEnum, OrientacaoSexualEnum, RacaCorEnum, SexoEnum, SituacaoMercadoEnum
from .residence_schema import ResidenciaResponse



class ResidentBase(BaseModel):
    residence_id: int

    # Identificação
    cns: Optional[str] = None
    cpf: Optional[str] = None
    nome_completo: str
    nome_social: Optional[str] = None
    data_nascimento: date
    sexo: SexoEnum
    raca_cor: RacaCorEnum
    etnia: Optional[str] = None
    nis_pis_pasep: Optional[str] = None

    # Filiação
    nome_mae: Optional[str] = None
    mae_desconhecida: bool = False
    nome_pai: Optional[str] = None
    pai_desconhecido: bool = False

    # Nacionalidade
    nacionalidade: NacionalidadeEnum
    pais_nascimento: Optional[str] = None
    dt_naturalizacao: Optional[date] = None
    portaria_naturalizacao: Optional[str] = None
    municipio_nascimento: Optional[str] = None

    # Contato
    telefone_celular: Optional[str] = None
    email: Optional[EmailStr] = None

    # Responsabilidade familiar
    eh_responsavel_familiar: bool = False
    cns_cpf_responsavel: Optional[str] = None
    relacao_parentesco: Optional[GrauParentescoEnum] = None

    # Sociodemográfico
    ocupacao: Optional[str] = None
    frequenta_escola: bool = False
    escolaridade: Optional[EscolaridadeEnum] = None
    situacao_mercado: Optional[SituacaoMercadoEnum] = None

    # Crianças 0–9 anos
    responsavel_crianca: Optional[List[str]] = None

    frequenta_cuidador: bool = False
    participa_grupos_comunitarios: bool = False
    possui_plano_saude: bool = False

    membro_povo_tradicional: bool = False
    povo_tradicional_desc: Optional[str] = None

    # Identidade e deficiência
    orientacao_sexual: Optional[OrientacaoSexualEnum] = None
    identidade_genero: Optional[IdentidadeGeneroEnum] = None

    possui_deficiencia: bool = False
    deficiencias_tipo: Optional[List[str]] = None

    # Saída do cadastro
    saida_cadastro: bool = False
    motivo_saida: Optional[str] = None
    data_obito: Optional[date] = None
    numero_do: Optional[str] = None

    @field_validator('*', mode='before')
    @classmethod
    def empty_string_to_none(cls, v: Any, info: ValidationInfo) -> Any:
        """
        Transforma strings vazias ('') vindas do React em None
        para não quebrar campos opcionais de Data, Email ou Enum.
        """
        if v == "":
            return None
        return v
    
    
class ResidentCreate(ResidentBase):
    pass



class ResidentUpdate(BaseModel):
    residence_id: Optional[int] = None

    cns: Optional[str] = None
    cpf: Optional[str] = None
    nome_completo: Optional[str] = None
    nome_social: Optional[str] = None
    data_nascimento: Optional[date] = None
    sexo: Optional[SexoEnum] = None
    raca_cor: Optional[RacaCorEnum] = None
    etnia: Optional[str] = None
    nis_pis_pasep: Optional[str] = None

    nome_mae: Optional[str] = None
    mae_desconhecida: Optional[bool] = None
    nome_pai: Optional[str] = None
    pai_desconhecido: Optional[bool] = None

    nacionalidade: Optional[NacionalidadeEnum] = None
    pais_nascimento: Optional[str] = None
    dt_naturalizacao: Optional[date] = None
    portaria_naturalizacao: Optional[str] = None
    municipio_nascimento: Optional[str] = None

    telefone_celular: Optional[str] = None
    email: Optional[EmailStr] = None

    eh_responsavel_familiar: Optional[bool] = None
    cns_cpf_responsavel: Optional[str] = None
    relacao_parentesco: Optional[GrauParentescoEnum] = None

    ocupacao: Optional[str] = None
    frequenta_escola: Optional[bool] = None
    escolaridade: Optional[EscolaridadeEnum] = None
    situacao_mercado: Optional[SituacaoMercadoEnum] = None

    responsavel_crianca: Optional[List[str]] = None
    frequenta_cuidador: Optional[bool] = None
    participa_grupos_comunitarios: Optional[bool] = None
    possui_plano_saude: Optional[bool] = None

    membro_povo_tradicional: Optional[bool] = None
    povo_tradicional_desc: Optional[str] = None

    orientacao_sexual: Optional[OrientacaoSexualEnum] = None
    identidade_genero: Optional[IdentidadeGeneroEnum] = None

    possui_deficiencia: Optional[bool] = None
    deficiencias_tipo: Optional[List[str]] = None

    saida_cadastro: Optional[bool] = None
    motivo_saida: Optional[str] = None
    data_obito: Optional[date] = None
    numero_do: Optional[str] = None




class ResidentResponse(ResidentBase):
    id: str


    class Config:
        from_attributes = True

class Resident(ResidentResponse):
    pass

